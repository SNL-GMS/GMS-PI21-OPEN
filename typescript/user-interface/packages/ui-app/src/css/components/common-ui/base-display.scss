.base-display {
  display: flex;
  position: relative;
  flex-direction: column;
  overflow: hidden;
  width: 100%;
  contain: strict; // Performance optimization to speed up render step
  &.full-width-height {
    height: 100%;
    width: 100%;
    flex-grow: 1;
  }
  &.scroll-box--x {
    overflow-x: auto;
    overflow-y: hidden;
  }
  &.scroll-box--y {
    overflow-x: hidden;
    overflow-y: auto;
  }
}
