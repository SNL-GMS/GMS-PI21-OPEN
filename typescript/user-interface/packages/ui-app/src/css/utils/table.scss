@mixin tableHeaderFontStyle {
  font-size: 16px;
  font-weight: 500;
  letter-spacing: 1px;
  padding: 0 10px;
}

.table-header-cell {
  @include tableHeaderFontStyle;
  color: var(--gms-main, $gms-main);

  &.ag-header-cell {
    padding: 0 10px;
  }
}

.table-group-header-cell {
  @include tableHeaderFontStyle;
  color: var(--gms-main, $gms-main);

  &.ag-header-cell {
    justify-content: center;
  }
}

.with-separated-rows-color .ag-row {
  border-top: 2px solid var(--gms-background, $gms-background);
  border-bottom: 2px solid var(--gms-background, $gms-background);
  border-color: var(--gms-background, $gms-background);

  .ag-cell {
    padding: 0;
  }
}

.force-left-justification > .table-cell .table-cell__value {
  justify-content: flex-start;
}
