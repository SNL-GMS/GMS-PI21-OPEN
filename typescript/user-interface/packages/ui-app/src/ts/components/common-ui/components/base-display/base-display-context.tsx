import type GoldenLayout from '@gms/golden-layout';
import React from 'react';

/**
 * Used for providing access to the base display size and a reference to the
 * golden layout container.
 */
export interface BaseDisplayContextData {
  /**
   * A reference to the golden layout container
   */
  glContainer: GoldenLayout.Container;

  /**
   * the width of the display in px, inclusive of the padding on the sides
   */
  widthPx: number;

  /**
   * the height of the display in px, inclusive of the padding on the sides
   */
  heightPx: number;
}

/**
 * Instantiate the Base Display Context and set up the defaults as undefined.
 */
export const BaseDisplayContext: React.Context<BaseDisplayContextData> = React.createContext<
  BaseDisplayContextData
>(undefined);
