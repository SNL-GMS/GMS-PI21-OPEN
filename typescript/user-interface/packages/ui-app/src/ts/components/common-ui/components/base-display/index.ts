export * from './base-display';
export * from './base-display-context';
