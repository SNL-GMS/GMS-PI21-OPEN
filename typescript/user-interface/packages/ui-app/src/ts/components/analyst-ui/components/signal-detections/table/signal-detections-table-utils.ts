import type { CommonTypes } from '@gms/common-model';
import { EventTypes, SignalDetectionTypes } from '@gms/common-model';
import { formatTimeForDisplay } from '@gms/common-util';
import type { AgGridCommunity, RowNode } from '@gms/ui-core-components';
import type {
  DisplayedSignalDetectionConfigurationEnum,
  EventStatus,
  SignalDetectionFetchResult
} from '@gms/ui-state';
import { SignalDetectionColumn } from '@gms/ui-state';
import type { AgGridReact } from 'ag-grid-react';
import Immutable from 'immutable';

import { SignalDetectionUtils } from '~analyst-ui/common/utils';
import { getSignalDetectionAssociationStatus } from '~analyst-ui/common/utils/event-util';
import {
  findAmplitudeFeatureMeasurementValue,
  findAzimuthFeatureMeasurementValue,
  findEmergenceAngleFeatureMeasurementValue,
  findFeatureMeasurementChannelName,
  findLongPeriodFirstMotionFeatureMeasurementValue,
  findRectilinearityFeatureMeasurementValue,
  findShortPeriodFirstMotionFeatureMeasurementValue,
  findSlownessFeatureMeasurementValue
} from '~analyst-ui/common/utils/signal-detection-util';
import { EdgeTypes } from '~analyst-ui/components/events/types';
import type { SignalDetectionRow } from '~analyst-ui/components/signal-detections/types';
import { messageConfig } from '~analyst-ui/config/message-config';
import {
  formatNumberForDisplayFixedThreeDecimalPlaces,
  getTableCellStringValue,
  setRowNodeSelection
} from '~common-ui/common/table-utils';

import {
  nonIdealStateLoadingSignalDetections,
  nonIdealStateNoSignalDetections,
  nonIdealStateNoSignalDetectionsErrorState,
  nonIdealStateNoSignalDetectionsSyncedTimeRange
} from '../signal-detections-non-ideal-states';

/**
 * Determines which non-ideal state, if any, should be presented.
 *
 * @returns NonIdealState, or undefined if not needed.
 */
export const handleNonIdealState = (
  signalDetectionsQuery: SignalDetectionFetchResult,
  isSynced: boolean
): JSX.Element | undefined => {
  if (!!signalDetectionsQuery.pending || !signalDetectionsQuery.data)
    return nonIdealStateLoadingSignalDetections;

  if (
    !signalDetectionsQuery.isLoading &&
    !signalDetectionsQuery.isError &&
    !signalDetectionsQuery.pending &&
    signalDetectionsQuery.data?.length === 0
  )
    return isSynced
      ? nonIdealStateNoSignalDetectionsSyncedTimeRange
      : nonIdealStateNoSignalDetections;

  if (signalDetectionsQuery.isError && signalDetectionsQuery.data?.length === 0)
    return nonIdealStateNoSignalDetectionsErrorState;

  return undefined;
};

/**
 * Given a timeRange, determines if the provided arrivalTime falls within that range
 * Returns corresponding {@link edgeType}
 *
 * @param timeRange if not defined, return Unknown
 * @param arrivalTime if not defined, return Unknown
 */
export function getEdgeType(
  timeRange: CommonTypes.TimeRange,
  arrivalTime: number
): string | EdgeTypes {
  if (
    (!arrivalTime && arrivalTime !== 0) ||
    !timeRange ||
    (!timeRange?.startTimeSecs && timeRange.startTimeSecs !== 0) ||
    (!timeRange?.endTimeSecs && timeRange.endTimeSecs !== 0)
  )
    return messageConfig.invalidCellText;
  if (arrivalTime < timeRange.startTimeSecs) return EdgeTypes.BEFORE;
  if (arrivalTime > timeRange.endTimeSecs) return EdgeTypes.AFTER;
  return EdgeTypes.INTERVAL;
}

/**
 * in charge of setting the row class that dims edge event rows
 */
export const edgeSDRowClassRules: AgGridCommunity.RowClassRules = {
  'edge-SD-row': params => params.data.edgeType !== EdgeTypes.INTERVAL
};

/**
 * This is the set of default columns to be displayed in the SD table
 * This object gets updated when columns are selected/deselected in the column picker so that state
 * doesn't get lost
 */
export const signalDetectionsColumnsToDisplay: Immutable.Map<
  SignalDetectionColumn,
  boolean
> = Immutable.Map([
  ...Object.values(SignalDetectionColumn).map<[SignalDetectionColumn, boolean]>(v => [v, true]),
  [SignalDetectionColumn.phaseConfidence, false],
  [SignalDetectionColumn.rectilinearity, false],
  [SignalDetectionColumn.emergenceAngle, false],
  [SignalDetectionColumn.shortPeriodFirstMotion, false],
  [SignalDetectionColumn.longPeriodFirstMotion, false]
]);

/**
 * These two fields have in-band error codes, if we see a '-1', we need to instead display "Unknown"
 * these fields are otherwise treated as the rest of the numeric fields
 *
 * @param value
 */
export function formatRectilinearityOrEmergenceForDisplay(value: number): string {
  if (value === -1) return 'Unknown';
  return formatNumberForDisplayFixedThreeDecimalPlaces(value);
}

/**
 * Called by ag-grid to determine if an external filter is present/active.
 */
export function agGridIsExternalFilterPresent(
  signalDetectionFilterState: Record<DisplayedSignalDetectionConfigurationEnum, boolean>
): boolean {
  const {
    signalDetectionUnassociated,
    signalDetectionAssociatedToOpenEvent,
    signalDetectionAssociatedToCompletedEvent,
    signalDetectionAssociatedToOtherEvent,
    signalDetectionBeforeInterval,
    signalDetectionAfterInterval
  } = signalDetectionFilterState;

  // If any of the event associations are unchecked then the table must be filtered.
  return (
    !signalDetectionUnassociated ||
    !signalDetectionAssociatedToOpenEvent ||
    !signalDetectionAssociatedToCompletedEvent ||
    !signalDetectionAssociatedToOtherEvent ||
    !signalDetectionBeforeInterval ||
    !signalDetectionAfterInterval
  );
}

/**
 * Passed to ag-grid. Should return true if external filter passes, otherwise false.
 */
export function agGridDoesExternalFilterPass(
  node: RowNode,
  signalDetectionFilterState: Record<DisplayedSignalDetectionConfigurationEnum, boolean>
): boolean {
  const {
    signalDetectionUnassociated,
    signalDetectionAssociatedToOpenEvent,
    signalDetectionAssociatedToCompletedEvent,
    signalDetectionAssociatedToOtherEvent,
    signalDetectionBeforeInterval,
    signalDetectionAfterInterval
  } = signalDetectionFilterState;

  if (!signalDetectionAfterInterval && node.data.edgeType === EdgeTypes.AFTER) {
    return false;
  }
  if (!signalDetectionBeforeInterval && node.data.edgeType === EdgeTypes.BEFORE) {
    return false;
  }

  switch ((node.data as SignalDetectionRow).assocStatus) {
    case EventTypes.AssociationStatus.COMPLETE_ASSOCIATED:
      return signalDetectionAssociatedToCompletedEvent;
    case EventTypes.AssociationStatus.OPEN_ASSOCIATED:
      return signalDetectionAssociatedToOpenEvent;
    case EventTypes.AssociationStatus.OTHER_ASSOCIATED:
      return signalDetectionAssociatedToOtherEvent;
    case EventTypes.AssociationStatus.UNASSOCIATED:
      return signalDetectionUnassociated;
    default:
      return true;
  }
}

/**
 * Builds a single Signal detection row given a signalDetection object
 *
 * @param sd Signal detection data for which the row is built
 * @param events Used to query for association status
 * @param openEventId Used to query for association status
 * @param eventsStatuses Used to query for association status
 * @param timeRange Used to determine if SD is an edge SD or not
 */
export function buildSignalDetectionRow(
  sd: SignalDetectionTypes.SignalDetection,
  events: EventTypes.Event[],
  eventsStatuses: Record<string, EventStatus>,
  openEventId: string,
  timeRange: CommonTypes.TimeRange
): SignalDetectionRow {
  const featureMeasurements = SignalDetectionTypes.Util.getCurrentHypothesis(
    sd.signalDetectionHypotheses
  )?.featureMeasurements;
  return {
    id: sd.id,
    unsavedChanges: 'TBD',
    assocStatus: getSignalDetectionAssociationStatus(sd, events, openEventId, eventsStatuses),
    conflict: 'TBD',
    station: getTableCellStringValue(sd?.station.name),
    channel: findFeatureMeasurementChannelName(featureMeasurements),
    phase: getTableCellStringValue(
      SignalDetectionUtils.findPhaseFeatureMeasurementValue(featureMeasurements)?.value
    ),
    phaseConfidence: formatNumberForDisplayFixedThreeDecimalPlaces(
      SignalDetectionUtils.findPhaseFeatureMeasurementValue(featureMeasurements)?.confidence
    ),
    time: formatTimeForDisplay(
      SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(featureMeasurements)
        ?.arrivalTime?.value
    ),
    timeStandardDeviation: formatNumberForDisplayFixedThreeDecimalPlaces(
      SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(featureMeasurements)
        ?.arrivalTime?.standardDeviation
    ),
    azimuth: formatNumberForDisplayFixedThreeDecimalPlaces(
      findAzimuthFeatureMeasurementValue(featureMeasurements)?.measuredValue?.value
    ),
    azimuthStandardDeviation: formatNumberForDisplayFixedThreeDecimalPlaces(
      findAzimuthFeatureMeasurementValue(featureMeasurements)?.measuredValue?.standardDeviation
    ),
    slowness: formatNumberForDisplayFixedThreeDecimalPlaces(
      findSlownessFeatureMeasurementValue(featureMeasurements)?.measuredValue?.value
    ),
    slownessStandardDeviation: formatNumberForDisplayFixedThreeDecimalPlaces(
      findSlownessFeatureMeasurementValue(featureMeasurements)?.measuredValue?.standardDeviation
    ),
    amplitude: formatNumberForDisplayFixedThreeDecimalPlaces(
      findAmplitudeFeatureMeasurementValue(
        featureMeasurements,
        SignalDetectionTypes.FeatureMeasurementType.AMPLITUDE_A5_OVER_2
      )?.amplitude?.value
    ),
    period: formatNumberForDisplayFixedThreeDecimalPlaces(
      findAmplitudeFeatureMeasurementValue(
        featureMeasurements,
        SignalDetectionTypes.FeatureMeasurementType.AMPLITUDE_A5_OVER_2
      )?.period
    ),
    sNR: formatNumberForDisplayFixedThreeDecimalPlaces(
      SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurement(featureMeasurements)?.snr?.value
    ),
    rectilinearity: formatRectilinearityOrEmergenceForDisplay(
      findRectilinearityFeatureMeasurementValue(featureMeasurements)?.measuredValue?.value
    ),
    emergenceAngle: formatRectilinearityOrEmergenceForDisplay(
      findEmergenceAngleFeatureMeasurementValue(featureMeasurements)?.measuredValue?.value
    ),
    shortPeriodFirstMotion: getTableCellStringValue(
      findShortPeriodFirstMotionFeatureMeasurementValue(featureMeasurements)?.value
    ),
    longPeriodFirstMotion: getTableCellStringValue(
      findLongPeriodFirstMotionFeatureMeasurementValue(featureMeasurements)?.value
    ),
    rejected: SignalDetectionTypes.Util.getCurrentHypothesis(sd?.signalDetectionHypotheses).rejected
      ? 'True'
      : 'False',
    edgeType: getEdgeType(
      timeRange,
      SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(featureMeasurements)
        ?.arrivalTime?.value
    )
  };
}

/**
 * Given an array of signal detections, builds rows for the SD table
 * returns an empty array when given null/empty/undefined input
 *
 * @param signalDetections Data is used as the basis for each row
 * @param events Used to query for association status
 * @param openEventId Used to query for association status
 * @param eventsStatuses Used to query for association status
 * @param timeRange Used to determine if SD is an edge SD or not
 */
export function buildSignalDetectionRows(
  signalDetections: SignalDetectionTypes.SignalDetection[],
  events: EventTypes.Event[],
  eventsStatuses: Record<string, EventStatus>,
  openEventId: string,
  timeRange: CommonTypes.TimeRange
): SignalDetectionRow[] {
  if (!signalDetections || signalDetections.length === 0) return [];

  return signalDetections.map(sd =>
    buildSignalDetectionRow(sd, events, eventsStatuses, openEventId, timeRange)
  );
}

/**
 * Given a table ref, checks to make sure the table exists and then updates the displayed columns
 * to match {@param columnsToDisplay}
 *
 * @param tableRef
 * @param columnsToDisplay
 */
export function updateColumns(
  tableRef: React.MutableRefObject<AgGridReact>,
  columnsToDisplay: Immutable.Map<SignalDetectionColumn, boolean>
): Immutable.Map<SignalDetectionColumn, boolean> | undefined {
  if (tableRef?.current?.columnApi) {
    columnsToDisplay.forEach((shouldDisplay, columnName) => {
      tableRef.current?.columnApi.setColumnVisible(columnName, shouldDisplay);
    });
    return columnsToDisplay;
  }
  return undefined;
}

/**
 * Cycles through all rows in the table and updates their row selection attribute
 * Sets rowSelection to true if their row id is in {@param selectedSdIds}, false otherwise
 *
 * @param tableRef
 * @param selectedSdIds
 */
export function updateRowSelection(
  tableRef: React.MutableRefObject<AgGridReact>,
  selectedSdIds: string[]
): React.MutableRefObject<AgGridReact> {
  tableRef.current.api.forEachNode(node =>
    setRowNodeSelection(node, selectedSdIds.includes(node.id))
  );
  return tableRef;
}
