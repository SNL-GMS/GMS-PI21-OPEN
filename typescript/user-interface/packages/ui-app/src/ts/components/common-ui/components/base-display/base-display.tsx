/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/prop-types */
import { getDataAttributesFromProps, useElementSize, useForceUpdate } from '@gms/ui-util';
import defer from 'lodash/defer';
import * as React from 'react';

import { TabContextMenu } from '~analyst-ui/common/menus/tab-context-menu';

import type { BaseDisplayContextData } from './base-display-context';
import { BaseDisplayContext } from './base-display-context';
import type { BaseDisplayProps } from './types';

const useForceUpdateOnFirstRender = () => {
  const forceUpdate = useForceUpdate();
  React.useEffect(() => {
    defer(forceUpdate);
    // !FIX ESLINT Validate and check REACT HOOK dependencies
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
};

/**
 * A base display that should be at the base of all display components.
 * Adds consistent padding to each display, and exposes the width and height
 * of the display in the BaseDisplayContext.
 *
 * @param props requires a reference to the glContainer.
 * Also accepts data attributes in the form 'data-cy': 'example-component'
 */
export const BaseDisplay: React.FunctionComponent<React.PropsWithChildren<
  BaseDisplayProps
  // eslint-disable-next-line react/function-component-definition
>> = props => {
  /**
   * Base display size behavior
   */
  const [displayRef, heightPx, widthPx] = useElementSize();

  /**
   * On the very first mount, call forceUpdate so that height and width will
   * propagate to the consumers of the context.
   */
  useForceUpdateOnFirstRender();

  /**
   * the context menu handler, if provided
   */
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { onContextMenu, tabName } = props;

  /**
   * Get any data attributes provided to this display (like data-cy attributes)
   */
  const dataAttributes = getDataAttributesFromProps(props);

  const baseDisplayContextData: BaseDisplayContextData = React.useMemo(
    () => ({
      glContainer: props.glContainer,
      widthPx: props.glContainer?.width ?? widthPx ?? 0,
      heightPx: props.glContainer?.height ?? heightPx ?? 0
    }),
    [heightPx, props.glContainer, widthPx]
  );

  return (
    <div
      className={`base-display ${props.className ?? ''}`}
      ref={ref => {
        displayRef.current = ref;
      }}
      onContextMenu={onContextMenu}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...dataAttributes}
    >
      <BaseDisplayContext.Provider value={baseDisplayContextData}>
        {tabName ? <TabContextMenu tabName={tabName} /> : null}
        {props.children}
      </BaseDisplayContext.Provider>
    </div>
  );
};
