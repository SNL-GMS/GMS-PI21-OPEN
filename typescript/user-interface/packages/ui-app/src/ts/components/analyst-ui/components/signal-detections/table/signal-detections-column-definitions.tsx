import { EventTypes } from '@gms/common-model';
import type { AgGridCommunity, ColumnDefinition } from '@gms/ui-core-components';
import { SignalDetectionColumn } from '@gms/ui-state';

import { getAssocStatusString } from '~analyst-ui/common/utils/signal-detection-util';
import { signalDetectionsColumnsToDisplay } from '~analyst-ui/components/signal-detections/table/signal-detections-table-utils';
import type { SignalDetectionRow } from '~analyst-ui/components/signal-detections/types';
import { signalDetectionColumnDisplayStrings } from '~analyst-ui/components/signal-detections/types';
import { messageConfig } from '~analyst-ui/config/message-config';
import { largeCellWidthPx, medCellWidthPx } from '~common-ui/common/table-types';
import { caseInsensitiveComparator, numericStringComparator } from '~common-ui/common/table-utils';
import { unsavedChangesIcon } from '~resources/unsaved-changes-icon';

import { DirtyDotCellRenderer } from './signal-detections-dirty-dot';

const standardDeviationHeaderWidthPx = 130;

/* applies classes to table cells conditionally based on content, used in this case for styling event association cells */
const eventAssociationCellClassRules: AgGridCommunity.CellClassRules = {
  'sd-cell--open-associated': params =>
    params.value === EventTypes.AssociationStatus.OPEN_ASSOCIATED,
  'sd-cell--complete-associated': params =>
    params.value === EventTypes.AssociationStatus.COMPLETE_ASSOCIATED,
  'sd-cell--other-associated': params =>
    params.value === EventTypes.AssociationStatus.OTHER_ASSOCIATED,
  'sd-cell--unassociated': params => params.value === EventTypes.AssociationStatus.UNASSOCIATED
};

function SDAssociationStatusTooltipValueGetter(params) {
  const tooltip = getAssocStatusString(params.value);
  if (tooltip === messageConfig.invalidCellText) return params.value;
  return tooltip;
}

export const signalDetectionsColumnDefs: ColumnDefinition<
  SignalDetectionRow,
  unknown,
  unknown,
  unknown,
  unknown
>[] = [
  {
    headerComponentParams: {
      template: unsavedChangesIcon
    },
    field: SignalDetectionColumn.unsavedChanges,
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.unsavedChanges),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.unsavedChanges),
    cellRendererFramework: DirtyDotCellRenderer,
    width: 30
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.assocStatus),
    field: SignalDetectionColumn.assocStatus,
    headerTooltip: 'Event association status',
    cellClassRules: eventAssociationCellClassRules,
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.assocStatus),
    tooltipValueGetter: params => SDAssociationStatusTooltipValueGetter(params),
    width: 83
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.conflict),
    field: SignalDetectionColumn.conflict,
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.conflict),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.conflict)
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.station),
    field: SignalDetectionColumn.station,
    cellClass: 'monospace',
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.station),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.station)
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.channel),
    field: SignalDetectionColumn.channel,
    cellClass: 'monospace',
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.channel),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.channel),
    width: medCellWidthPx
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.phase),
    field: SignalDetectionColumn.phase,
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.phase),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.phase),
    comparator: caseInsensitiveComparator
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.phaseConfidence),
    field: SignalDetectionColumn.phaseConfidence,
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.phaseConfidence),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.phaseConfidence)
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.time),
    field: SignalDetectionColumn.time,
    headerTooltip: 'Arrival time',
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.time),
    width: largeCellWidthPx,
    initialSort: 'asc'
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(
      SignalDetectionColumn.timeStandardDeviation
    ),
    field: SignalDetectionColumn.timeStandardDeviation,
    headerTooltip: 'Arrival time standard deviation',
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.timeStandardDeviation)
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.azimuth),
    field: SignalDetectionColumn.azimuth,
    headerTooltip: 'Observed receiver-to-source azimuth',
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.azimuth),
    comparator: numericStringComparator
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(
      SignalDetectionColumn.azimuthStandardDeviation
    ),
    field: SignalDetectionColumn.azimuthStandardDeviation,
    headerTooltip: 'Azimuth standard deviation',
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.azimuthStandardDeviation),
    comparator: numericStringComparator,
    width: standardDeviationHeaderWidthPx
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.slowness),
    field: SignalDetectionColumn.slowness,
    headerTooltip: 'Observed slowness',
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.slowness),
    comparator: numericStringComparator
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(
      SignalDetectionColumn.slownessStandardDeviation
    ),
    field: SignalDetectionColumn.slownessStandardDeviation,
    headerTooltip: 'Slowness standard deviation',
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.slownessStandardDeviation),
    comparator: numericStringComparator,
    width: standardDeviationHeaderWidthPx
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.amplitude),
    field: SignalDetectionColumn.amplitude,
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.amplitude),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.amplitude),
    comparator: numericStringComparator
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.period),
    field: SignalDetectionColumn.period,
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.period),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.period),
    comparator: numericStringComparator
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.sNR),
    field: SignalDetectionColumn.sNR,
    headerTooltip: 'Arrival time signal to noise ratio',
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.sNR),
    comparator: numericStringComparator
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.rectilinearity),
    field: SignalDetectionColumn.rectilinearity,
    headerTooltip: 'Signal rectilinearity',
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.rectilinearity),
    comparator: numericStringComparator
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.emergenceAngle),
    field: SignalDetectionColumn.emergenceAngle,
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.emergenceAngle),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.emergenceAngle),
    comparator: numericStringComparator
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(
      SignalDetectionColumn.shortPeriodFirstMotion
    ),
    field: SignalDetectionColumn.shortPeriodFirstMotion,
    headerTooltip: signalDetectionColumnDisplayStrings.get(
      SignalDetectionColumn.shortPeriodFirstMotion
    ),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.shortPeriodFirstMotion),
    width: medCellWidthPx
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(
      SignalDetectionColumn.longPeriodFirstMotion
    ),
    field: SignalDetectionColumn.longPeriodFirstMotion,
    headerTooltip: signalDetectionColumnDisplayStrings.get(
      SignalDetectionColumn.longPeriodFirstMotion
    ),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.longPeriodFirstMotion),
    width: medCellWidthPx
  },
  {
    headerName: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.rejected),
    field: SignalDetectionColumn.rejected,
    headerTooltip: signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.rejected),
    hide: !signalDetectionsColumnsToDisplay.get(SignalDetectionColumn.rejected)
  }
];
