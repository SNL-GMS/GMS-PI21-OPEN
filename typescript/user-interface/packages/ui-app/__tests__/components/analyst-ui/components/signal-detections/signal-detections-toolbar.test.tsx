/* eslint-disable react/jsx-no-constructed-context-values */
import { getStore } from '@gms/ui-state';
import { render } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';

import { signalDetectionsColumnsToDisplay } from '../../../../../src/ts/components/analyst-ui/components/signal-detections/table/signal-detections-table-utils';
import {
  onChangeHandler,
  SignalDetectionsToolbar
} from '../../../../../src/ts/components/analyst-ui/components/signal-detections/toolbar/signal-detections-toolbar';
import { BaseDisplayContext } from '../../../../../src/ts/components/common-ui/components/base-display/base-display-context';

describe('Signal Detection Toolbar', () => {
  test('can be mounted', () => {
    expect(SignalDetectionsToolbar).toBeDefined();
  });

  test('should match snapshot', () => {
    const { container } = render(
      <Provider store={getStore()}>
        <BaseDisplayContext.Provider
          value={{
            glContainer: {} as any,
            widthPx: 100,
            heightPx: 100
          }}
        >
          <SignalDetectionsToolbar
            setSelectedSDColumnsToDisplay={jest.fn()}
            selectedSDColumnsToDisplay={signalDetectionsColumnsToDisplay}
          />
        </BaseDisplayContext.Provider>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('checkbox list onChangeHandler with configuration arguments', () => {
    const dispatch = jest.fn();
    const eventObject = {
      syncWaveform: false,
      signalDetectionBeforeInterval: true,
      signalDetectionAfterInterval: true,
      signalDetectionAssociatedToOpenEvent: true,
      signalDetectionAssociatedToCompletedEvent: true,
      signalDetectionAssociatedToOtherEvent: true,
      signalDetectionUnassociated: false
    };
    onChangeHandler(dispatch)(eventObject);
    expect(dispatch).toHaveBeenCalledTimes(1);
  });
});
