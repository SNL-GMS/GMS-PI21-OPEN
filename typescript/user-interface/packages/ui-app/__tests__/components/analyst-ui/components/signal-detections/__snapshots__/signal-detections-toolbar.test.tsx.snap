// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signal Detection Toolbar should match snapshot 1`] = `
<div>
  <div
    class="toolbar"
    style="width: 60px;"
  >
    <div
      class="toolbar__left-group"
    >
      <div
        class="toolbar-item toolbar-item__left"
      >
        <div>
          <div
            data-cy="sd-table-detections-toggle"
          >
            <button
              class="bp4-button bp4-align-left toolbar-button"
              title="Display signal detections within the queried workflow time range."
              type="button"
            >
              <span
                class="bp4-button-text"
              >
                <span>
                  Show detections
                </span>
                <span
                  aria-hidden="true"
                  class="bp4-icon bp4-icon-chevron-down"
                  icon="chevron-down"
                >
                  <svg
                    data-icon="chevron-down"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <path
                      d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="toolbar-item toolbar-item__left"
      >
        <div>
          <div
            data-cy="sd-table-column-picker"
          >
            <button
              class="bp4-button bp4-align-left toolbar-button"
              title="Select columns to be shown in the signal detection table below"
              type="button"
            >
              <span
                class="bp4-button-text"
              >
                <span>
                  Show columns
                </span>
                <span
                  aria-hidden="true"
                  class="bp4-icon bp4-icon-chevron-down"
                  icon="chevron-down"
                >
                  <svg
                    data-icon="chevron-down"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <path
                      d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="toolbar__center-group"
    >
      <div
        class="toolbar__whitespace"
        style="width: 60px;"
      />
    </div>
    <div
      class="toolbar__right-group"
    />
  </div>
</div>
`;
